components/service/glean/src/main/java/mozilla/components/service/glean/storages/GenericScalarStorageEngine.kt

```diff
@@ -0,0 +1,203 @@
+     */
+    protected fun serializeSingleMetric(value: ScalarType): String? {
+        return value.toString()
+    }
+
+    /**
+     * Deserialize the metrics with a lifetime = User that are on disk.
+     * This will be called the first time a metric is used or before a snapshot is
+     * taken.
+     *
+     * @return A [SharedPreferences] reference that will be used to inititialize [userLifetimeStorage]
+     */
+    @Suppress("TooGenericExceptionCaught")
+    open fun deserializeUserLifetime(): SharedPreferences {
+        val prefs =
+            applicationContext.getSharedPreferences(this.javaClass.simpleName, Context.MODE_PRIVATE)
+
+        val metrics = try {
+            prefs.all.entries
+        } catch (e: NullPointerException) {
+            // If we fail to deserialize, we can log the problem but keep on going.
+            logger.error("Failed to deserialize metric with 'user' lifetime")
+            return prefs
+        }
+
+        for ((metricStoragePath, metricValue) in metrics) {
+            if (!metricStoragePath.contains('#')) {
+                continue
+            }
+
+            // Split the stored name in 2: we expect it to be in the format
+            // store#metric.name
+            val (storeName, metricName) =
+                metricStoragePath.split('#', limit = 2)
+            if (storeName.length <= 1) {
+                continue
+            }
+
+            val storeData = dataStores[Lifetime.User.ordinal].getOrPut(storeName) { mutableMapOf() }
+            // Only set the stored value if we're able to deserialize the persisted data.
+            deserializeSingleMetric(metricValue)?.let {
+                storeData[metricName] = it
+            } ?: logger.warn("Failed to deserialize $metricStoragePath")
+        }
+
+        return prefs
+    }
+
+    /**
+     * Retrieves the [recorded metric data][ScalarType] for the provided
+     * store name.
+     *
+     * Please note that the [Lifetime.Application] lifetime is handled implicitly
+     * by never clearing its data. It will naturally clear out when restarting the
+     * application.
+     *
+     * @param storeName the name of the desired store
+     * @param clearStore whether or not to clear the requested store. Not that only
+     *        metrics stored with a lifetime of [Lifetime.Ping] will be cleared.
+     *
+     * @return the [ScalarType] recorded in the requested store
+     */
+    @Synchronized
+    fun getSnapshot(storeName: String, clearStore: Boolean): GenericDataStorage<ScalarType>? {
+        val allLifetimes: GenericDataStorage<ScalarType> = mutableMapOf()
+
+        // Make sure data with "user" lifetime is loaded before getting the snapshot.
+        // We still need to catch exceptions here, as `getAll()` might throw.
+        @Suppress("TooGenericExceptionCaught")
+        try {
+            userLifetimeStorage.all
+        } catch (e: NullPointerException) {
```

Comment:
Is `NullPointerException` the most generic one here? (I have no clue about Kotlin, so just trying to understand it)

Reply:
There's the Java `RuntimeException`, which this one inherits from. However, I'm only interested in catching this one, since this is the only one that `all` can throw.